File: src/backend/lineup_model.py
"""Pure-Python data model for a lineup event.

This module defines plain dataclasses that capture the *complete* state of a
lineup — titles, timestamps, slots, open-decks config, and DJ metadata — with
**zero** dependency on Tkinter or any GUI framework.

The model is the single source of truth.  UI widgets read from / write to
instance attributes, and the ``EventBus`` broadcasts changes so other
subsystems (output generation, auto-save, etc.) can react without coupling.

Typical flow
------------
1. User types in a ``CTkEntry`` → UI handler calls ``model.set_title("…")``.
2. The setter publishes ``"model_changed"`` on the bus.
3. ``OutputGenerator`` (subscribed) rebuilds the preview text.
4. The output panel (subscribed) writes the new text to the textbox.
"""

from __future__ import annotations

import datetime
from dataclasses import dataclass, field
from typing import TYPE_CHECKING

from .types import DJInfo, EventSnapshot, SlotData

if TYPE_CHECKING:
    from .event_bus import EventBus


# ── Re-export for backward compatibility ──────────────────────────────────
__all__ = ["SlotData", "DJInfo", "EventSnapshot", "LineupModel"]


# ── Live mutable model ────────────────────────────────────────────────────

class LineupModel:
    """The runtime model that backs the editor UI.

    Every mutating setter publishes ``"model_changed"`` on the bus so
    subscribers (output panel, auto-save timer, etc.) can react.

    The model is deliberately *not* a dataclass itself because it manages
    change-notification logic and provides ``snapshot()`` for consumers
    that need a frozen view.
    """

    def __init__(self, bus: EventBus) -> None:
        self._bus = bus

        self.title: str = ""
        self.vol: str = ""
        self.timestamp: str = ""
        self.master_duration: int = 60
        self.genres: list[str] = []
        self.slots: list[SlotData] = []
        self.names_only: bool = False
        self.output_format: str = "discord"
        self.saved_djs: list[DJInfo] = []

    # ── Snapshot ──────────────────────────────────────────────────────

    def snapshot(self) -> EventSnapshot:
        """Return a frozen copy suitable for ``OutputGenerator``."""
        return EventSnapshot(
            title=self.title,
            vol=self.vol,
            timestamp=self.timestamp,
            genres=list(self.genres),
            slots=[SlotData(s.name, s.genre, s.duration) for s in self.slots],
            names_only=self.names_only,
            output_format=self.output_format,
            saved_djs=[DJInfo(d.name, d.stream, d.exact_link) for d in self.saved_djs],
        )

    # ── Bulk load (e.g. loading a saved event) ────────────────────────

    def load_from_dict(self, data: dict) -> None:
        """Populate the model from a plain dict (event save format)."""
        self.title = data.get("title", "")
        self.vol = data.get("vol", "")
        self.timestamp = data.get("timestamp", "")
        self.master_duration = int(data.get("master_duration", 60))
        self.genres = list(data.get("genres", []))
        self.names_only = bool(data.get("names_only", False))
        self.slots = [
            SlotData(
                name=s.get("name", ""),
                genre=s.get("genre", ""),
                duration=int(s.get("duration", 60)),
            )
            for s in data.get("slots", [])
        ]
        self._notify()

    def to_dict(self) -> dict:
        """Serialize to the same plain-dict format used for persistence."""
        return {
            "title": self.title,
            "vol": self.vol,
            "timestamp": self.timestamp,
            "master_duration": str(self.master_duration),
            "genres": list(self.genres),
            "names_only": self.names_only,
            "slots": [
                {"name": s.name, "genre": s.genre, "duration": str(s.duration)}
                for s in self.slots
            ],
        }

    # ── Change notification ───────────────────────────────────────────

    def notify(self) -> None:
        """Publicly trigger a model-changed broadcast.

        Useful when the UI has mutated a ``SlotData`` in-place and needs
        to tell subscribers about it.
        """
        self._notify()

    def _notify(self) -> None:
        self._bus.publish("model_changed")

File: web/src/models/lineup-model.js
/**
 * LineupModel — pure JS port of desktop/src/backend/models/lineup_model.py
 *
 * Mutable runtime model. Every setter publishes "model_changed" on the bus
 * so subscribers (output panel, auto-save, etc.) can react.
 */

export class SlotData {
  constructor({ name = "", genre = "", duration = 60 } = {}) {
    this.name = name;
    this.genre = genre;
    this.duration = Number(duration) || 60;
  }

  clone() {
    return new SlotData({ name: this.name, genre: this.genre, duration: this.duration });
  }

  toObject() {
    return { name: this.name, genre: this.genre, duration: this.duration };
  }
}

export class DJInfo {
  constructor({ name = "", stream = "", exactLink, exact_link } = {}) {
    this.name = name;
    this.stream = stream;
    this.exactLink = Boolean(exactLink ?? exact_link ?? false);
  }

  clone() {
    return new DJInfo({ name: this.name, stream: this.stream, exactLink: this.exactLink });
  }

  toObject() {
    return { name: this.name, stream: this.stream, exactLink: this.exactLink };
  }
}

/**
 * An immutable snapshot of the full event state — passed to OutputGenerator.
 */
export class EventSnapshot {
  constructor({
    title = "",
    vol = "",
    groupName = "",
    collab = false,
    collabWith = "",
    timestamp = "",
    genres = [],
    slots = [],
    namesOnly = false,
    outputFormat = "discord",
    streamLinkFormat = "",
    savedDjs = [],
    socialLinks = {},
    discordRoleMentions = [],
  } = {}) {
    this.title = title;
    this.vol = vol;
    this.groupName = groupName;
    this.collab = collab;
    this.collabWith = collabWith;
    this.timestamp = timestamp;
    this.genres = [...genres];
    this.slots = slots.map((s) => (s instanceof SlotData ? s.clone() : new SlotData(s)));
    this.namesOnly = namesOnly;
    this.outputFormat = outputFormat;
    this.streamLinkFormat = streamLinkFormat;
    this.savedDjs = savedDjs.map((d) => (d instanceof DJInfo ? d.clone() : new DJInfo(d)));
    this.socialLinks = { ...socialLinks };
    this.discordRoleMentions = [...discordRoleMentions];
  }

  get startDatetime() {
    if (!this.timestamp) return new Date();
    const d = new Date(this.timestamp.replace(" ", "T"));
    return isNaN(d.getTime()) ? new Date() : d;
  }

  get fullTitle() {
    let base = /^\d+$/.test(this.vol) ? `${this.title} VOL.${this.vol}` : this.title;
    if (this.collab && this.collabWith) {
      base += ` x ${this.collabWith}`;
    }
    return base;
  }
}

/**
 * The mutable runtime model backing the editor UI.
 */
export class LineupModel {
  constructor(bus) {
    this._bus = bus;

    this.title = "";
    this.vol = "";
    this.groupName = "";
    this.collab = false;
    this.collabWith = "";
    this.timestamp = "";
    this.genres = [];
    this.slots = [];
    this.namesOnly = false;
    this.outputFormat = "discord";
    this.streamLinkFormat = ""; // "" | "quest" | "pc"
    this.savedDjs = [];
    this.socialLinks = {};
    this.discordRoleMentions = []; // [{id, name}]
  }

  // ── Snapshot ──────────────────────────────────────────────────────────

  snapshot() {
    return new EventSnapshot({
      title: this.title,
      vol: this.vol,
      groupName: this.groupName,
      collab: this.collab,
      collabWith: this.collabWith,
      timestamp: this.timestamp,
      genres: [...this.genres],
      slots: this.slots.map((s) => s instanceof SlotData ? s.clone() : new SlotData(s)),
      namesOnly: this.namesOnly,
      outputFormat: this.outputFormat,
      streamLinkFormat: this.streamLinkFormat,
      savedDjs: this.savedDjs.map((d) => d instanceof DJInfo ? d.clone() : new DJInfo(d)),
      socialLinks: { ...this.socialLinks },
      discordRoleMentions: this.discordRoleMentions.map((r) => ({ ...r })),
    });
  }

  // ── Serialization ──────────────────────────────────────────────────────

  toObject() {
    return {
      title: this.title,
      vol: this.vol,
      groupName: this.groupName,
      collab: this.collab,
      collabWith: this.collabWith,
      timestamp: this.timestamp,
      genres: [...this.genres],
      slots: this.slots.map((s) => s.toObject()),
      namesOnly: this.namesOnly,
      outputFormat: this.outputFormat,
      streamLinkFormat: this.streamLinkFormat,
      socialLinks: { ...this.socialLinks },
      discordRoleMentions: this.discordRoleMentions.map((r) => ({ ...r })),
    };
  }

  loadFromObject(data) {
    this.title = data.title ?? "";
    this.vol = data.vol ?? "";
    this.groupName = data.groupName ?? "";
    this.collab = Boolean(data.collab);
    this.collabWith = data.collabWith ?? "";
    this.timestamp = data.timestamp ?? "";
    this.genres = Array.isArray(data.genres) ? [...data.genres] : [];
    this.slots = Array.isArray(data.slots)
      ? data.slots.map((s) => new SlotData(s))
      : [];
    this.namesOnly = Boolean(data.namesOnly);
    this.outputFormat = data.outputFormat ?? "discord";
    this.streamLinkFormat = data.streamLinkFormat ?? "";
    this.socialLinks = data.socialLinks ? { ...data.socialLinks } : {};
    this.discordRoleMentions = Array.isArray(data.discordRoleMentions)
      ? data.discordRoleMentions.map((r) => ({ ...r }))
      : [];
    this._notify();
  }

  clear() {
    this.loadFromObject({});
  }

  // ── Change notification ───────────────────────────────────────────────

  notify() {
    this._notify();
  }

  _notify() {
    this._bus.publish("model_changed");
  }
}
